### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by Tutuchan**
This PR adds a lot of documentation for existing functions in the package. 

It also includes a few classes definitions and code changes (most notably replacing `line` and `character` parameters by `position`) in order to stick closer to the LSP Specification.
 
I plan on continuing this documentation process and I'm open to discuss any ideas or changes you can think of on this front. I also started working on the textDocument/definition request implementation in another branch.

Please let me know if I can do anything more !

### CODE REVIEW

**Comment by randy3k** [comment]
Awesome. I have also invited you to the organization, please check email.

As a rule of thumb, please do not push directly to master. I would like to give my 2cents before merging anything.

Great job, keep ahead.

Meanwhile, I am looking into 
https://microsoft.github.io/debug-adapter-protocol/

**Comment by randy3k** [comment]
I guess you are a vscode user, judging from the `.vscode` directory. You may also be interested in joining the discussion at https://github.com/Ikuyadeu/vscode-R/issues/98
about the future of Vscode-R.

**Comment by Tutuchan** [comment]
Thanks for the invitation and the link to the vscode-R repo. I'll remember not pushing to master.

First order of business is finishing the documentation for the existing functions then doing some refactoring.

File: R/workspace.R [hunk 1]
```diff
@@ -74,41 +106,41 @@ Workspace <- R6::R6Class("Workspace",
             }
         },
 
-        guess_package = function(object) {
+        guess_package = function(funct) {
```
Comment (randy3k) [inline R/workspace.R]: this should really be `object`, objects are not limited to functions.

File: R/utils.R [hunk 2]
```diff
@@ -21,41 +34,83 @@ find_package <- function(path = getwd()) {
...
+        if (Sys.time() - last_execution_time > t) {
+            last_execution_time <<- Sys.time()
             fun(...)
         }
     }
 }
 
+#' sanitize package objects names
+#'
+#' Remove unwanted objects, _e.g._ objects whose name starts with `_` or `.`.
```
Comment (randy3k) [inline R/utils.R]: actually we still allow name starts with `.`, it is used mostly for operators like
`%>%` and assignments like `names<-`.

File: R/utils.R [hunk 3]
```diff
@@ -21,41 +34,83 @@ find_package <- function(path = getwd()) {
...
 }
 
+#' get the parent process pid
+#'
+#' @keywords internal
 getppid <- function() {
     .Call("do_getppid", PACKAGE = "languageserver")
 }
 
+#' delay a function execution
```
Comment (randy3k) [inline R/utils.R]: actually it doesn't delay. It will simply ignore the call if last execution time is less than `t`.

Comment (randy3k) [reply R/utils.R]: `leisurize` may not be a good name, though I cannot think of a good one.

Comment (Tutuchan) [reply R/utils.R]: `throttle` maybe ?

**Comment by Tutuchan** [comment]
Thanks, I'll check them out and update tomorrow.

**Comment by Tutuchan** [comment]
I pushed some fixes based on you comments. I'd like to add more documentation before we merge this PR, I'll let you know when I think it is ready.

**Comment by Tutuchan** [comment]
@randy3k I think this branch is ready for the time being. We can add more documentation or refine it in the future if you'd like.

After the merge, I'll stop working on my fork and work directly on branches of this repo if that's alright.

**Comment by randy3k** [comment]
I am running some cran checks and there are some warnings

```
checking Rd \usage sections ... WARNING
Undocumented arguments in documentation object 'completion_reply'
  ‘id’ ‘uri’ ‘workspace’ ‘document’ ‘position’

Undocumented arguments in documentation object 'hover_reply'
  ‘id’ ‘uri’ ‘workspace’ ‘document’ ‘position’

Undocumented arguments in documentation object 'on_initialize'
  ‘self’ ‘id’ ‘params’

Undocumented arguments in documentation object 'signature_reply'
  ‘id’ ‘uri’ ‘workspace’ ‘document’ ‘position’
```

It is due to the use of nested `@template` tag https://github.com/REditorSupport/languageserver/blob/975f3edd58b770bef11a579c32c98bb206147777/man-roxygen/reply-parameters.R#L1-L5. AKAIK, the template tag cannot be used inside `man-roxygen` to avoid infinite loop.

**Comment by Tutuchan** [comment]
My bad, I did not run the CRAN check. I'm checking this out.

**Comment by randy3k** [comment]
It was also my responsibility to setup a travis-ci check.

**Pull request merged**